### DISCUSSION TITLE
[Plugin] How to parameterize?

### DISCUSSION THREAD

**Discussion by deeplook** [abhinavsingh/proxy.py#519] [Q&A] [answered]
I'm looking for ways to parameterize custom plugins. In my case I want to write one that acts like a "man in the middle" who some value to the HTTP header before passing on the request. Is there some recommended pattern? So far I have something like this, which is not ideal:

```python
from typing import Optional
from proxy.http.parser import HttpParser
from proxy.http.proxy import HttpProxyBasePlugin

class ModifyHeaderPlugin(HttpProxyBasePlugin):
    """Modify request header before sending to upstream server."""

    # This seems not to be a working pattern:
    # def __init__(self, headers: Optional[dict] = None, **kwargs):
    #     self.added_headers = headers or {}
    #     # HttpProxyBasePlugin.__init__(self, **kwargs)

    def before_upstream_connection(self, request: HttpParser) -> Optional[HttpParser]:
        return request

    def handle_client_request(self, request: HttpParser) -> Optional[HttpParser]:
        # I'd prefer to pass this token to the constructor:
        request.add_header(b'Authorization', bytes(f"Bearer {token}", "utf-8"))
        return request

    def handle_upstream_chunk(self, chunk: memoryview) -> memoryview:
        return chunk

    def on_upstream_connection_close(self) -> None:
        pass
```

**Comment by abhinavsingh** [comment]
@deeplook Apologies first of all for missing all your discussions.  I'll get to your other threads shortly.

`proxy.py` plugins can have their own flags.  See https://github.com/abhinavsingh/proxy.py/blob/develop/proxy/plugin/filter_by_client_ip.py#L20. Let me know if this doesn't help.

**Comment by abhinavsingh** [comment, marked as answer]
> 
> ```python
>
>     def handle_client_request(self, request: HttpParser) -> Optional[HttpParser]:
>         # I'd prefer to pass this token to the constructor:
>         request.add_header(b'Authorization', bytes(f"Bearer {token}", "utf-8"))
>         return request
> ```

If the token here is static i.e. same for all outgoing requests, then `flags` will simply do the trick.  You can define a flag like:

```
flags.add_argument(
    '--bearer-token',
    type=str,
    default=None,
    help='Default: None.  Bearer token.'
)
```

Then you can use this flag within the plugin as:

`self.flags.bearer_token`

If the token itself is dynamic, then it will be best to dynamically fetch the token from within the plugin when necessary.

I hope this helps.

**Comment by deeplook** [comment]
Thanks for your response! I've looked at https://mitmproxy.org in the meantime, too, but then realized, I might not need one at all in my case. ;)
